ctdcal/io.py

```diff
@@ -34,6 +56,99 @@ def load_exchange_btl(btl_file: Union[str, Path]) -> pd.DataFrame:
     )
 
 
+def load_exchange_ctd(
+    ctd_file: Union[str, Path, BufferedIOBase],
+    n_files=None,
+    recursed=False,
+) -> pd.DataFrame:
+    """
+    Load WHP-exchange CTD file(s) (_ct1.csv) into DataFrame. File(s) can be on local
+    file system or downloaded from an appropriate cchdo.ucsd.edu link
+    (e.g., https://cchdo.ucsd.edu/data/19434/325020210316_ct1.zip)
+
+    Adapted from cchdo.hydro package.
+
+    Parameters
+    ----------
+    ctd_file : str or Path
+        Name or URL of file to be loaded
+
+    n_files : int, optional
+        Number of files to load from .zip archive
+
+    Returns
+    -------
+    header : dict or list of dict
+        File metadata from header(s) (e.g., EXPOCODE, STNNBR, CASTNO)
+    df : DataFrame or list of DataFrame
+        Loaded CTD file(s)
+    """
+    # read from url (.zip)
+    if isinstance(ctd_file, (str, Path)) and str(ctd_file).startswith("http"):
+        log.info("Loading CTD file from http link")
+        data_raw = BytesIO(requests.get(ctd_file).content)
+
+    # read from file
+    elif isinstance(ctd_file, (str, Path)):
+        log.info("Loading CTD file from local file")
+        with open(ctd_file, "rb") as f:
+            data_raw = BytesIO(f.read())
+
+    # read from open file
+    elif isinstance(ctd_file, BufferedIOBase):
+        log.info("Loading open file object")
+        data_raw = BytesIO(ctd_file.read())
+
+    # .zip special behavior
+    if is_zipfile(data_raw):
+        log.info("Loading CTD files from .zip")
+
+        if recursed is True:
+            raise ZipImportError("Recursive .zip files encountered... exiting")
+
+        data_raw.seek(0)  # is_zipfile moves cursor to EOF, reset to start
+        zip_contents = []
+        with ZipFile(data_raw) as zf:
+            for zipinfo in zf.infolist():
+                zip_contents.append(BytesIO(zf.read(zipinfo)))
+
+        # list comprehension is same as using functools.partial, just different syntax
+        return zip(
+            *[load_exchange_ctd(zc, recursed=True) for zc in zip_contents[:n_files]]
+        )
+
+    else:
+        data_raw.seek(0)  # is_zipfile moves cursor to EOF, reset to start
+        file = data_raw.read().decode("utf8").splitlines(keepends=True)
+
+    # process metadata
+    for idx, line in enumerate(file):
+        # find header info
+        if line.startswith("NUMBER_HEADERS"):
+            header_ind = idx
+
+        # find index of units row
+        if line.startswith("CTDPRS"):
```

Comment:
Careful of two things here:
1. ordering of columns is not guaranteed, just "preferred"
2. There can be spaces in front of the param name

Reply:
Would replacing this with 

```
if "CTDPRS" in line:  # for CTD

if "EXPOCODE" in line:  # for bottle
``` 

be sufficient? I'm pretty sure those are both required columns, right?

Reply:
Hm, this breaks if any comment line has either of those words in it... back to the drawing board.

Reply:
Resolved with e93c72d0cae57c70258cebe52f5e40b2d7426575

Reply:
Yeah I think that will do it.

The CCHDO code tends to actually use that NUMBER_HEADERS thing for alignment, we want our code to throw when it isn't correct (so we know to fix it).